src/csrf.test.js

```diff
@@ -0,0 +1,228 @@
+            Cookies.remove(myCookie);
+        });
+
+        beforeEach(function() {
+            sync = sinon.fake.returns(magicReturnValue);
+            syncCSRF = wrapWithCSRF(sync, myHeader, myCookie);
+            method = 'create';
+            model = {id: 1, url: path};
+            headers = {'x-requested-with': 'XMLHttpRequest'};
+            options = {success: _.noop};
+        });
+
+        function makePristineCopies() {
+            var _headers = _.clone(headers),
+                _options = _.clone(options);
+            if (_options && _options.headers) _options.headers = _headers;
+            return {headers: _headers, options: _options};
+        }
+
+        function assertBasicPassthrough(pristine) {
+            assert(sync.calledOnce);
+            var matchOptions = sinon.match(pristine.options);
+            assert(sync.calledWith(method, model, matchOptions));
+        }
+
+        function assertNoSideEffects(pristine) {
+            assert.deepStrictEqual(options, pristine.options);
+        }
+
+        function assertHeaderAdded() {
+            var passedOptions = sync.firstCall.args[2];
+            assert(passedOptions.headers[myHeader] === myToken);
+        }
+
+        function assertHeaderNotAdded() {
+            var passedOptions = sync.firstCall.args[2];
+            assert(passedOptions === options);
+            assert(
+                !passedOptions ||
+                !('headers' in passedOptions) ||
+                !(myHeader in passedOptions.headers) ||
+                passedOptions.headers[myHeader] !== myToken
+            );
+        }
+
+        function completeCheck(headerCheck) {
+            return function() {
+                var pristine = makePristineCopies.call(this);
+                var result = syncCSRF(method, model, options);
+                assertBasicPassthrough.call(this, pristine);
+                assert(result === magicReturnValue);
+                headerCheck.call(this);
+                assertNoSideEffects.call(this, pristine);
+            };
+        }
+
+        var assertTransparentPassthrough = completeCheck(assertHeaderNotAdded),
+            assertAugmentedPassthrough = completeCheck(assertHeaderAdded);
+
+        describe('defaults to transparent passthrough...', function() {
+            it('when no arguments are passed', function() {
+                method = undefined;
+                model = undefined;
+                options = undefined;
+                assertTransparentPassthrough.call(this);
+            });
+
+            it('when no url can be determined', function() {
+                delete model.url;
+                assertTransparentPassthrough.call(this);
+            });
+
+            it('when the method is "read", ...', function() {
+                method = 'read';
+                assertTransparentPassthrough.call(this);
+            });
+
+            it('regardless of whether model.url is function, ...', function() {
```

Comment:
It would be more readable if all `it` statements start with "when".

Reply:
I omitted "when" here because this spec belongs with the previous and I'm trying to express how it differs from it. I wasn't happy with this approach myself, though. If you know of a way to express that adjacent specs are related that makes more consistent spec titles, please let me know.